src/util/sighash.rs

```diff
@@ -0,0 +1,904 @@
+                    }
+                    sha256::Hash::from_engine(enc)
+                },
+            };
+            self.common_cache = Some(cache);
+        }
+        self.common_cache.as_ref().unwrap() // safe to unwrap because we checked is_none()
+    }
+
+    fn segwit_cache(&mut self) -> &SegwitCache {
+        if self.segwit_cache.is_none() {
+            let cache = SegwitCache {
+                prevouts: sha256d::Hash::from_inner(
+                    sha256::Hash::hash(&self.common_cache().prevouts).into_inner(),
+                ),
+                sequences: sha256d::Hash::from_inner(
+                    sha256::Hash::hash(&self.common_cache().sequences).into_inner(),
+                ),
+                outputs: sha256d::Hash::from_inner(
+                    sha256::Hash::hash(&self.common_cache().outputs).into_inner(),
+                ),
+            };
+            self.segwit_cache = Some(cache);
+        }
+        self.segwit_cache.as_ref().unwrap() // safe to unwrap because we checked is_none()
+    }
+
+    fn taproot_cache(&mut self, prevouts: &[TxOut]) -> &TaprootCache {
+        if self.taproot_cache.is_none() {
+            let mut enc_amounts = sha256::Hash::engine();
+            let mut enc_script_pubkeys = sha256::Hash::engine();
+            for prevout in prevouts {
+                prevout.value.consensus_encode(&mut enc_amounts).unwrap();
+                prevout
+                    .script_pubkey
+                    .consensus_encode(&mut enc_script_pubkeys)
+                    .unwrap();
+            }
+            let cache = TaprootCache {
+                amounts: sha256::Hash::from_engine(enc_amounts),
+                script_pubkeys: sha256::Hash::from_engine(enc_script_pubkeys),
+            };
+            self.taproot_cache = Some(cache);
+        }
+        self.taproot_cache.as_ref().unwrap() // safe to unwrap because we checked is_none()
+    }
+}
+
+impl<R: DerefMut<Target = Transaction>> SigHashCache<R> {
+    /// When the SigHashCache is initialized with a mutable reference to a transaction instead of a
+    /// regular reference, this method is available to allow modification to the witnesses.
+    ///
+    /// This allows in-line signing such as
+    /// ```
+    /// use bitcoin::blockdata::transaction::{Transaction, SigHashType};
+    /// use bitcoin::util::bip143::SigHashCache;
+    /// use bitcoin::Script;
+    ///
+    /// let mut tx_to_sign = Transaction { version: 2, lock_time: 0, input: Vec::new(), output: Vec::new() };
+    /// let input_count = tx_to_sign.input.len();
+    ///
+    /// let mut sig_hasher = SigHashCache::new(&mut tx_to_sign);
+    /// for inp in 0..input_count {
+    ///     let prevout_script = Script::new();
+    ///     let _sighash = sig_hasher.signature_hash(inp, &prevout_script, 42, SigHashType::All);
+    ///     // ... sign the sighash
+    ///     sig_hasher.access_witness(inp).push(Vec::new());
+    /// }
```

Comment:
Document that it panics if `input_index` is out of bounds?

Reply:
No need to comment anymore since now it returns an Option in f223be618fff7f06d6a1b2e0ca94c506f3ea2fd7 (commented the old method in `bip143.rs` though)